impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java

```diff
@@ -309,26 +317,112 @@ public synchronized void removeBroker(int nodeId) throws UnsupportedOperationExc
             kafkaContainer.stop();
         }
     }
 
+    @Override
+    public synchronized void restartBrokers(Predicate<Integer> nodeIdPredicate, boolean abruptShutdown) {
+        var kafkaContainersToRestart = brokers.entrySet().stream().filter(e -> nodeIdPredicate.test(e.getKey())).map(Map.Entry::getValue).toList();
+
+        var inspectCommands = kafkaContainersToRestart.stream().map(kc -> kc.getDockerClient().inspectContainerCmd(kc.getContainerId())).toList();
+
+        kafkaContainersToRestart.forEach(kc -> {
+            if (abruptShutdown) {
+                kc.stop();
+            }
+            else {
+                gracefulStop(kc);
+            }
+        });
+
+        // Await the existing containers going away.
+        inspectCommands.forEach(inspectContainer -> {
+            Utils.awaitCondition(Long.valueOf(STARTUP_TIMEOUT.toMillis()).intValue(), TimeUnit.MILLISECONDS).until(() -> {
+                try {
+                    inspectContainer.exec();
+                }
+                catch (NotFoundException e) {
+                    return true;
+                }
+                return false;
+            });
+        });
+
+        if (zookeeper != null) {
+            // In the zookeeper case, may as well wait for the session timeout. The replacement kafka broker
+            // won't be permitted to join until at least that interval has passed. This prevents excessive
+            // spinning during restart.
+            var config = clusterConfig.getBrokerConfigs(() -> this).findFirst();
+            var zkSessionTimeout = config.map(KafkaClusterConfig.ConfigHolder::getProperties).map(p -> p.getProperty(KafkaConfig.ZkSessionTimeoutMsProp(), "0"))
+                    .map(Long::parseLong);
+            zkSessionTimeout.ifPresent(
+                    timeOut -> {
+                        try {
+                            LOGGER.log(Level.DEBUG, "Awaiting zookeeper session timeout {0}ms so that the broker ephemeral nodes expire.", timeOut);
+                            Thread.sleep(timeOut);
+                        }
+                        catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                    });
+        }
+
+        LOGGER.log(Level.DEBUG, "Restarting {0}/{1} broker(s)", kafkaContainersToRestart.size(), getNumOfBrokers());
+        kafkaContainersToRestart.forEach(kc -> {
+            var originalStartupAttempts = kc.getStartupAttempts();
+            try {
+                // We need to control the restart loop ourselves. Unfortunately testcontainers does not
+                // allow a delay to be introduced between startup attempts and start-ups in a tight loop
+                // seems to lead to podman unreliability (at least on the Mac).
+                kc.withStartupAttempts(1);
+                Unreliables.retryUntilSuccess(Long.valueOf(STARTUP_TIMEOUT.toMillis()).intValue(), TimeUnit.MILLISECONDS,
```

Comment:
I went to check the `Unreliables` API looking for backoff options and noticed the [repo](https://github.com/rnorth/duct-tape) has been archived. Do we think its good enough or should we look at moving to awaitility or something else?